frontend/model/contracts/group.js

```diff
@@ -312,32 +312,138 @@ sbp('chelonia/defineContract', {
+            else return sortedPeriodKeys[sortedPeriodKeys.length - 2] ?? undefined
+          } else {
+            // If no period has been stored yet, then we're in the waiting period and can do arithmetic.
+            return dateToPeriodStamp(addTimeToDate(distributionDate, -distributionPeriodLength))
+          }
+        }
+        const index = sortedPeriodKeys.indexOf(periodStamp)
+        if (index === -1) {
+          // Maybe the given stamp is wrong and has no associated period,
+          // but is one period length ahead of a known one.
+          // TODO: just return 'undefined' when sure it's always safe.
+          const maybePreviousStamp = dateToPeriodStamp(
+            addTimeToDate(dateFromPeriodStamp(periodStamp), -distributionPeriodLength)
+          )
+          return sortedPeriodKeys.includes(maybePreviousStamp) ? maybePreviousStamp : undefined
+        }
+        // If index is 0 then the caller will have to check the archive.
+        if (index === 0) return
+        return sortedPeriodKeys[index - 1]
       }
     },
+    // May return 'undefined', in which case the caller should check archived data.
     periodAfterPeriod (state, getters) {
       return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), len))
+        if (!isPeriodStamp(periodStamp)) throw new TypeError('must be periodStamp')
+        const { distributionDate, distributionPeriodLength } = getters.groupSettings
+        if (!distributionDate) return
+        // This is not always the current period stamp.
+        const distributionDateStamp = dateToPeriodStamp(distributionDate)
+        const sortedPeriodKeys = getters.groupSortedPeriodKeys
+
+        // Maybe the distribution date has just been updated,
+        // and the corresponding period has not started or is not stored yet.
+        if (periodStamp > distributionDateStamp) return
+        if (periodStamp === distributionDateStamp) {
+          return dateToPeriodStamp(addTimeToDate(distributionDate, distributionPeriodLength))
+        }
+        // Maybe we're in the waiting period but haven't stored it yet.
+        if (!sortedPeriodKeys.length) {
+          const waitingPeriodStamp = dateToPeriodStamp(
+            addTimeToDate(distributionDate, -distributionPeriodLength)
+          )
+          return periodStamp === waitingPeriodStamp ? distributionDate : undefined
+        }
+        const index = sortedPeriodKeys.indexOf(periodStamp)
+        if (index === -1) return
+        // Maybe the given stamp is the last stored one but doesn't match the distribution date.
+        if (index === sortedPeriodKeys.length - 1) return dateToPeriodStamp(distributionDate)
+        // Now 'index + 1' is always a valid index.
+        return sortedPeriodKeys[index + 1]
```

Comment:
Similarly, this seems like it can be simplified to:

```js
    periodAfterPeriod (state, getters) {
      return (periodStamp: string) => {
        const { distributionPeriodLength } = getters.groupSettings
        const sortedPeriods = getters.groupSortedPeriodKeys
        let latestPeriod
        for (const i = sortedPeriods.length - 1; i >= 0; --i) {
          latestPeriod = sortedPeriods[i]
          if (periodStamp >= latestPeriod) break
        }
        const onePeriodAfterLatest = dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(latestPeriod), distributionPeriodLength))
        if (periodStamp >= onePeriodAfterLatest) {
          return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), distributionPeriodLength))
        }
        return latestPeriod
      }
    },
```